### PR TITLE
Return OOB errors from blockDeviceBackedFile.WriteAt

### PR DESCRIPTION
**Pull request by mostynb**

### CODE REVIEW

**Comment by EdSchouten** [comment]
Great find! Could you add some testing coverage for this as well?

File: pkg/filesystem/block_device_backed_file_pool_test.go [hunk 1]
```diff
@@ -309,4 +309,12 @@ func TestBlockDeviceBackedFilePool(t *testing.T) {
 		sectorAllocator.EXPECT().FreeList([]uint32{0, 0, 0, 0, 0, 0, 0, 0, 5})
 		require.NoError(t, f.Close())
 	})
+
+	t.Run("WriteAt", func(t *testing.T) {
+		f, err := pool.NewFile()
+		require.NoError(t, err)
+
+		_, err = f.WriteAt([]byte{0}, -1)
+		require.Error(t, err, "Expected an OOB error for negative offset")
```
Comment (EdSchouten) [inline pkg/filesystem/block_device_backed_file_pool_test.go:318]: Could you do a testutil.RequireEqualStatus() here instead? Just so we can make sure we got the string formatting of the error message right. Thanks!

Comment (mostynb) [reply pkg/filesystem/block_device_backed_file_pool_test.go]: Sure- I wasn't sure if you wanted to check for a specific error, but checking for the status is a good compromise.

File: go.mod [hunk 2]
```diff
@@ -8,11 +8,13 @@ require (
 	github.com/aws/aws-sdk-go v1.40.30
 	github.com/bazelbuild/remote-apis v0.0.0-20210812183132-3e816456ee28
 	github.com/buildbarn/bb-storage v0.0.0-20210824114323-bba43727809c
+	github.com/golang/mock v1.6.0 // indirect
```
Comment (EdSchouten) [inline go.mod:11]: Q: how did these suddenly appear? We didn’t introduce any new dependencies, right?

File: pkg/filesystem/block_device_backed_file_pool_test.go [hunk 3]
```diff
@@ -309,4 +310,12 @@ func TestBlockDeviceBackedFilePool(t *testing.T) {
 		sectorAllocator.EXPECT().FreeList([]uint32{0, 0, 0, 0, 0, 0, 0, 0, 5})
 		require.NoError(t, f.Close())
 	})
+
+	t.Run("WriteAt", func(t *testing.T) {
+		f, err := pool.NewFile()
+		require.NoError(t, err)
+
+		_, err = f.WriteAt([]byte{0}, -1)
+		testutil.RequireEqualStatus(t, status.Error(codes.InvalidArgument, ""), err)
```
Comment (EdSchouten) [inline pkg/filesystem/block_device_backed_file_pool_test.go:319]: Did this test pass like this? We aren’t supposed to return an empty message, right?

Comment (mostynb) [reply pkg/filesystem/block_device_backed_file_pool_test.go]: Oops- I assumed it was just comparing the code, updated.

File: go.mod (hunk 2 above)
Comment (mostynb) [reply go.mod:11]: These aren't new dependencies, I guess bazel fills these in somehow when testing? I was testing without bazel, `go test ./...` prints a bunch of errors about these being missing:

````
$ git diff go.mod 
$ go test ./...
go: github.com/buildbarn/bb-remote-execution/pkg/blobstore: package github.com/golang/mock/gomock imported from implicitly required module; to add missing requirements, run:
	go get github.com/golang/mock/gomock@v1.6.0
go: github.com/buildbarn/bb-remote-execution/pkg/blobstore: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
go: github.com/buildbarn/bb-remote-execution/pkg/builder: package github.com/golang/mock/gomock imported from implicitly required module; to add missing requirements, run:
	go get github.com/golang/mock/gomock@v1.6.0
go: github.com/buildbarn/bb-remote-execution/pkg/builder: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
go: github.com/buildbarn/bb-remote-execution/pkg/builder/initialsizeclass: package github.com/golang/mock/gomock imported from implicitly required module; to add missing requirements, run:
	go get github.com/golang/mock/gomock@v1.6.0
go: github.com/buildbarn/bb-remote-execution/pkg/builder/initialsizeclass: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
go: github.com/buildbarn/bb-remote-execution/pkg/builder/platform: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
go: github.com/buildbarn/bb-remote-execution/pkg/cas: package github.com/golang/mock/gomock imported from implicitly required module; to add missing requirements, run:
	go get github.com/golang/mock/gomock@v1.6.0
go: github.com/buildbarn/bb-remote-execution/pkg/cas: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
go: github.com/buildbarn/bb-remote-execution/pkg/cleaner: package github.com/golang/mock/gomock imported from implicitly required module; to add missing requirements, run:
	go get github.com/golang/mock/gomock@v1.6.0
go: github.com/buildbarn/bb-remote-execution/pkg/cleaner: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
go: github.com/buildbarn/bb-remote-execution/pkg/clock: package github.com/golang/mock/gomock imported from implicitly required module; to add missing requirements, run:
	go get github.com/golang/mock/gomock@v1.6.0
go: github.com/buildbarn/bb-remote-execution/pkg/clock: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
go: github.com/buildbarn/bb-remote-execution/pkg/cloud/aws: package github.com/golang/mock/gomock imported from implicitly required module; to add missing requirements, run:
	go get github.com/golang/mock/gomock@v1.6.0
go: github.com/buildbarn/bb-remote-execution/pkg/cloud/aws: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
go: github.com/buildbarn/bb-remote-execution/pkg/filesystem: package github.com/golang/mock/gomock imported from implicitly required module; to add missing requirements, run:
	go get github.com/golang/mock/gomock@v1.6.0
go: github.com/buildbarn/bb-remote-execution/pkg/filesystem: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
go: github.com/buildbarn/bb-remote-execution/pkg/filesystem/fuse: package github.com/golang/mock/gomock imported from implicitly required module; to add missing requirements, run:
	go get github.com/golang/mock/gomock@v1.6.0
go: github.com/buildbarn/bb-remote-execution/pkg/filesystem/fuse: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
go: github.com/buildbarn/bb-remote-execution/pkg/runner: package github.com/golang/mock/gomock imported from implicitly required module; to add missing requirements, run:
	go get github.com/golang/mock/gomock@v1.6.0
go: github.com/buildbarn/bb-remote-execution/pkg/runner: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
go: github.com/buildbarn/bb-remote-execution/pkg/sync: package github.com/stretchr/testify/require imported from implicitly required module; to add missing requirements, run:
	go get github.com/stretchr/testify/require@v1.7.0
$ go get github.com/stretchr/testify/require@v1.7.0
$ go get github.com/golang/mock/gomock@v1.6.0
$ git diff go.mod 
diff --git a/go.mod b/go.mod
index d9a5777..7cbe2e6 100644
--- a/go.mod
+++ b/go.mod
@@ -8,11 +8,13 @@ require (
        github.com/aws/aws-sdk-go v1.40.30
        github.com/bazelbuild/remote-apis v0.0.0-20210812183132-3e816456ee28
        github.com/buildbarn/bb-storage v0.0.0-20210824114323-bba43727809c
+       github.com/golang/mock v1.6.0 // indirect
        github.com/golang/protobuf v1.5.2
        github.com/google/uuid v1.3.0
        github.com/gorilla/mux v1.8.0
        github.com/hanwen/go-fuse/v2 v2.1.0
        github.com/prometheus/client_golang v1.11.0
+       github.com/stretchr/testify v1.7.0 // indirect
        go.opentelemetry.io/otel v1.0.0-RC2
        go.opentelemetry.io/otel/metric v0.20.0 // indirect
        go.opentelemetry.io/otel/trace v1.0.0-RC2
````

However, there is a still generated(?) file that breaks bare "go test":
````
$ go test ./...
../../go/pkg/mod/github.com/buildbarn/bb-storage@v0.0.0-20210824114323-bba43727809c/pkg/otel/active_spans_reporting_http_handler.go:28:13: pattern bootstrap.css: no matching files found
````

Comment (EdSchouten) [reply go.mod:11]: Interesting. I generated go.mod the other day by just running ‘go get -d ./…’ and it didn’t add those dependencies.

I’d say, just revert changes to this file as part of this PR. This is something to investigate separately. In addition to that. ‘go test’ is not supported for this code base. Just run ‘bazel test’.

Comment (mostynb) [reply go.mod]: OK, done.

**Comment by EdSchouten** [comment]
Thanks a lot for fixing this!

**Pull request merged**